web/modules/custom/search_api_field/src/Plugin/Field/FieldWidget/SearchWidget.php

```diff
@@ -390,6 +447,220 @@ protected function getFacets() {
+        'wrapper' => $wrapper_id,
+      ],
+      '#limit_validation_errors' => [
+        array_merge(
+          $parents,
+          [$field_name, $delta, 'wrapper', 'query_builder']
+        ),
+      ],
+    ];
+
+    return $element;
+  }
+
+  /**
+   * Ajax callback to re-render the query builder wrapper.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   *
+   * @return array
+   *   The form portion to render.
+   */
+  public function ajaxUpdateQueryBuilder(array $form, FormStateInterface $form_state) {
+    $button = $form_state->getTriggeringElement();
+    $wrapper_parents = array_slice($button['#array_parents'], 0, array_search('query_builder', $button['#array_parents'], TRUE));
+    $element = NestedArray::getValue($form, $wrapper_parents);
+
+    return $element['query_builder'];
+  }
+
+  /**
+   * Submit handler for the add filter button.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public function submitAddFilter(array $form, FormStateInterface $form_state) {
+    $button = $form_state->getTriggeringElement();
+    $wrapper = NestedArray::getValue($form, array_slice($button['#array_parents'], 0, -1));
+
+    // Extract field id and plugin id from the selected option. Since field ids
+    // cannot contain special characters, it's safe to explode on the first
+    // colon.
+    list($field_id, $plugin_id) = explode(':', $form_state->getValue($wrapper['field']['#parents']), 2);
+
+    // Extract element and widget elements.
+    $element = NestedArray::getValue($form, array_slice($button['#array_parents'], 0, -3));
+    $widget = NestedArray::getValue($form, array_slice($button['#array_parents'], 0, -4));
+
+    $field_name = $widget['#field_name'];
+    $parents = $widget['#field_parents'];
+    $field_state = static::getWidgetState($parents, $field_name, $form_state);
+    // Add the selected filter to the list for this specific delta.
+    $field_state['query_builder'][$element['#delta']]['filters'][] = [
+      'plugin' => $plugin_id,
+      'field' => $field_id,
+    ];
+    static::setWidgetState($parents, $field_name, $form_state, $field_state);
+
+    $form_state->setRebuild();
+  }
+
+  /**
+   * Submit handler for the remove filter button.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   */
+  public function submitRemoveFilter(array $form, FormStateInterface $form_state) {
+    $button = $form_state->getTriggeringElement();
+    $filter = NestedArray::getValue($form, array_slice($button['#array_parents'], 0, -1));
```

Comment:
Probably this is why the last entry is removed regardless of which "Remove filter" button is pressed.

Reply:
It should be caused by the wrong button being set as triggering element. The button name is not unique.